README.md

```diff
@@ -0,0 +1,560 @@
+#### Mutation
+ ```java
+ client.mutate(deleteTask)
+        .refetchQueries(tasksQuery)
+        .enqueue(new ApolloCall.Callback<DeleteTaskMutation.Data>() {
+```
+- We want to use `refetchQueries(myGetQuery)` in mutations just to refetch our get query so cache gets updated.
+
+#### Subscriptions
+   ```java
+   client.subscribe(deleteTaskSubscription)
+        .execute(new ApolloSubscriptionCall.Callback<DeleteTaskSubscription.Data>() {
+  ```
+- Currently the only way to refresh cache after receiving subscription response is to use `subscribeToMore`, however, this is not available in java.
+  After request has been sent we either receive `onResponse`, if request was successful, or `onFailure` if request has failed.
+
+ ```java
+  @Override
+public void onResponse(@NotNull Response<AllTasksQuery.Data> response) {
+    ON RESPONSE LOGIC
+}
+ @Override
+public void onFailure(@NotNull ApolloException e) {
+    ON FAILURE LOGIC
+}
+```
+### 2. Keycloak implementation
+To implement Keycloak with our app we have used [AppAuth](https://github.com/openid/AppAuth-Android). 
+You will need a keycloak instance running either on OpenShift or you can setup locally on Ionic Showcase server that has been used in our example app.
+
+- First step is to fetch well knows configuration for Keycloak
+
+```java
+mAuthService = new AuthorizationService(this);
+final AuthorizationServiceConfiguration.RetrieveConfigurationCallback retrieveCallback =
+        (authorizationServiceConfiguration, e) -> {
+            if (e != null) {
+                System.out.println("Failed to retrieve configuration for " + mobileService.getKIssuer());
+  } else {
+                makeAuthRequest(authorizationServiceConfiguration);
+  }
+        };
+
+String discoveryEndpoint = mobileService.getKIssuer() + ".well-known/openid-configuration";
+AuthorizationServiceConfiguration.fetchFromUrl(Uri.parse(discoveryEndpoint), retrieveCallback);
+```
+- Once we have received configuration from Keycloak we can build Authorization Request
+
+```java
+String k_REDIRECT_URI = "com.m.androidnativeapp:/oauth2redirect";
+
+AuthorizationRequest.Builder authorizationRequest = new AuthorizationRequest.Builder(
+        serviceConfiguration,
+  mobileService.getKClientId(),
+  ResponseTypeValues.CODE,
+  Uri.parse(k_REDIRECT_URI));
+
+AuthorizationRequest authRequest = authorizationRequest.build();
+
+Intent authIntent = mAuthService.getAuthorizationRequestIntent(authRequest);
+startActivityForResult(authIntent, RC_AUTH);
+```
+- It is important to note that we need to provide valid `redirect_URI` and `clientId`, in our sample app we are using `mobile-services.json` file that has been generated in [MDC](https://access.redhat.com/documentation/en-us/red_hat_mobile_developer_services/1/pdf/getting_started/Red_Hat_Mobile_Developer_Services-1-Getting_Started-en-US.pdf). Once authorization request has been send app is redirected to Keycloak log in screen.
```

Comment:
Also I think AppAuth has some docs explaining how to set `redirect_uri`. Probably should link it here.